packages/mui-material/src/styles/variants.d.ts

```diff
@@ -3,7 +3,9 @@ import { ComponentsPropsList } from './props';
 
 export type ComponentsVariants<Theme = unknown> = {
   [Name in keyof ComponentsPropsList]?: Array<{
-    props: Partial<ComponentsPropsList[Name]>;
+    props:
+      | Partial<ComponentsPropsList[Name]>
+      | ((props: Partial<ComponentsPropsList[Name]>) => boolean);
```

Comment:
I really wish we have `ownerState` instead of `props`:

```js
variants: [
  { ownerState: { … }, style: { … } },
  { ownerState: ({ color, variant }) => …, style: { … } },
]
```